### PR TITLE
PR #4 - Implemented Pagination

### PR DESCRIPTION
**Pull request by gdavidss**
## Description

This pull request features an implementation of pagination
Future pull request will involve local storage

## Milestones
This feature allows users to load thousands of messages without overconsumption of memory.

## Demo
Kapture 2022-08-01 at 12 02 10

### CODE REVIEW

File: ProjectVoyage/View Controllers/Chat message/MessagesViewController.m [hunk 1]
```diff
@@ -39,8 +39,10 @@ @interface MessagesViewController() <InputbarDelegate, UITableViewDataSource, UI
 // This assumes that there's only one other recipient in each chat, should be changed if groups are allowed
 @property (nonatomic, strong) PFUser *otherRecipient;
 
+// pagination variables
 @property (nonatomic, assign) NSInteger currentPageNumber;
 @property (nonatomic, assign) bool shouldKeepScrolling;
+@property (nonatomic, assign) int MESSAGES_PER_PAGE;
```
Comment (tliao33) [inline ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:45]: Feel like this should be managed as a const instead + renamed to use camel case

File: ProjectVoyage/View Controllers/Chat message/MessagesViewController.m [hunk 2]
```diff
@@ -113,18 +117,19 @@ - (void) liveQueryMessage {
...
+    // Using live query to immediately show the change
     self.liveQueryClient = [[PFLiveQueryClient alloc] initWithServer:@"wss://chat2markdown.b4a.io" applicationId:app_id clientKey:client_id];
     PFQuery *messageQuery = [PFQuery queryWithClassName:@"Message"];
     self.subscription = [self.liveQueryClient subscribeToQuery:messageQuery];
    
    __unsafe_unretained typeof(self) weakSelf = self;
    [self.subscription addUpdateHandler:^(PFQuery<PFObject *> * _Nonnull query, PFObject * _Nonnull object) {
        __strong typeof (self) strongSelf = weakSelf;
-       if (object) {
+       if (object){
```
Comment (tliao33) [inline ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:128]: nit: space

File: ProjectVoyage/View Controllers/Chat message/MessagesViewController.m [hunk 3]
```diff
@@ -143,46 +148,62 @@ - (Message *) findMessageByObjectId:(NSString *)objectId {
...
             [strongSelf->_tableView reloadData];
+            // GD The method below should only be called when the chat is first opened but...
+            // the problem is that in viewDidLoad the messages haven't been loaded yet
+            // because this thread hasn't finished loading the messages.
+            
+            //[strongSelf scrollToBottom];
         } else {
-            // GD throw alert
-            NSLog(@"%@", error.localizedDescription);
+            strongSelf->_shouldKeepScrolling = NO;
```
Comment (tliao33) [inline ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:184]: nit but I feel like naming something like canKeepScrolling would be more clear

File: ProjectVoyage/View Controllers/Chat message/MessagesViewController.m [hunk 4]
```diff
@@ -366,21 +386,24 @@ - (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger
...
-    // Open chat in the very last message
-    if (numberOfRows) {
-        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:(self.chat.messages.count - 1) inSection:0];
-        
-        [_tableView scrollToRowAtIndexPath:indexPath
-                    atScrollPosition:UITableViewScrollPositionBottom
-                    animated:animated];
+- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
+    if (_shouldKeepScrolling) {
+        if (scrollView.contentOffset.y + _tableView.safeAreaInsets.top == 0) {
```
Comment (tliao33) [inline ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:404]: nit: save the sum as a variable just to better explain what this check is doing?

**Review by tliao33 — COMMENTED** [review]
Left some nits but overall looks good. @kotojo to complete review
Also good job cleaning up this PR -- really like seeing the demo too!

**Review by tliao33 — APPROVED** [review]

**Comment by rainmaker490** [comment]
@gdavidss - what about when a user is going loading old messages & a chat participant sends messages at the same time. what is the behavior then?

so:
iOS App #1 - start pagination by scrolling
iOS App #2 - user sends a message to this thread while #1 is in process of scrolling

File: ProjectVoyage/View Controllers/Chat message/MessagesViewController.m [hunk 5]
```diff
@@ -143,46 +148,62 @@ - (Message *) findMessageByObjectId:(NSString *)objectId {
...
     __weak __typeof(self) weakSelf = self;
-    [query findObjectsInBackgroundWithBlock:^(NSArray *messages, NSError *error) {
+    [query findObjectsInBackgroundWithBlock:^(NSArray *reversedMessages, NSError *error) {
         __strong __typeof(weakSelf) strongSelf = weakSelf;
         if (!strongSelf) return;
-        if (messages != nil) {
-            strongSelf->_chat.messages = [messages mutableCopy];
+        if (reversedMessages == nil) {
+            NSLog(@"%@", error.localizedDescription);
+        } else if ([reversedMessages count] > 0) {
```
Comment (rainmaker490) [inline ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:169]: I think this should be:

if reversedMessages.count <= MESSAGES_PER_PAGE

Comment (rainmaker490) [reply ProjectVoyage/View Controllers/Chat message/MessagesViewController.m:169]: actually, this also works :)

**Review by kotojo — APPROVED** [review]

**Pull request merged**